jaxrs-spec/src/main/asciidoc/chapters/providers/_lifecycle_and_environment.adoc

```diff
@@ -33,6 +33,14 @@ discover provider classes via mechanisms such as class scanning. A
 JAX-RS implementation that supports automatic discovery of classes MUST
 process only those classes that are annotated with `@Provider`.
 
+[[services]]
+==== Services
+On runtime initialization, an implementation MUST discover and register certain service providers in `jakarta.ws.rs.core.Configuration` runtime configurations.
+For each service X, where X is `jakarta.ws.rs.core.Feature` or `jakarta.ws.rs.container.DynamicFeature`, the set of providers to be registered MUST be determined as follows:
+
+* first, if `ServiceLoader.load(Application.class)` returns no results, register all providers discovered with `ServiceLoader.load(X.class)`
+* additionally, if `ServiceLoader.load(Application.class, X.class.getClassLoader())` returns no results, register all providers discovered with `ServiceLoader.load(X.class, X.class.getClassLoader())`
```

Comment:
We should not use the `ServiceLoader` to determine the existence of an `Application` subclass.  This would only work if the JAX-RS archive contained a `META-INF/services/jakarta.ws.rs.core.Application` file that specified a subclass (or if there was a Java 11+ module that specified the application subclass).  Instead, I think this line should just mention the existence of an `Application` subclass - but it should also specify that the services should be registered if `Application`'s `getClasses` and `getSingletons` methods return null or empty sets. Many JAX-RS apps contains an `Application` subclass solely for the purpose of specifying an `@ApplicationPath` or providing additional properties, etc.  

Maybe we could change these lines to something like:
```suggestion
If an `Application` subclass exists in the application, and that subclass returns a non-empty collection for either the `getClasses` or `getSingletons` method, then the implementation MUST NOT register any providers specified via the service loader mechanism. If no `Application` subclass exists, or if the subclass exists but does not specify the set of classes or singletons to use for the application, then the implementation MUST register all providers discovered with `ServiceLoader.load(X.class)`.
```

Reply:
Thanks, those are good points. I've added a note about what it's for. The spec itself is updated in a slightly different way as suggested though, to allow for the case
> when essentially portable features are provided via jar files that are put in an internal server path (e.g. glassfish/modules in Payara).

dixit [arjantijms](https://github.com/eclipse-ee4j/jaxrs-api/pull/751#issuecomment-477816397)

What do you think of this version?

Reply:
@andymc12 I'm a bit confused by your last suggestion (and the new changes). If my application depends on REST library "Foo" and my application includes an `Application` subclass that returns non-empty collections from at least one of its method, how can library "Foo" extend the Jakarta REST runtime to function properly? I may have missed some earlier discussion about this.

Reply:
@spericas My thinking was that if the `Application` subclass returns a non-empty collection of providers, then we would not want to add any additional providers - even those that are part of the extension JARs.  The original text excluded the providers in the extension JARs if an `Application` subclass was present - so my comment was that some apps will contain an `Application` subclass but only to provide the `@ApplicationPath` and not to specify a strict collection of providers - in those cases, I think we want the providers in the extension JARs to be auto-registered.

Reply:
I'm not convinced that implementing `getClasses` or `getSingletons` should disable extension discovery for the classloader of the `Application` subclass. 

And to be honest: I'm not sure if we really need a way to opt out extensions from beeing processed at all. If an extension is on your classpath, why not enabing it? If somebody puts an extension into an location where it is visible to all applications deployed to an app server or puts it as a library into an EAR where it is visible to all contained modules, it is hopefully because of some good reason. Also, the same potential "problem" exists for CDI extensions as well. At least I don't know of any mechanism to opt out a CDI extension which is on the classpath of the app.

In my optionion it should be the responsibility of the extension developer to provide way to disable the extension although it is visible to a JAX-RS application. And this is actually super easy to do:

```java
public class MyExtension implements Feature {

  @Override
  public boolean configure( FeatureContext context ) {
    if( context.getConfiguration().getProperty( "foo.bar.disable" ) != null ) {
      context.register( MyProvider1.class );
      context.register( MyProvider2.class );
      return true;
    }
    return false;
  }

}
```

So IMO extensions should always be processes, whether or not `getClasses` or `getSingletons` is implemented.

Reply:
Look at this case: An application explicitly implements `getClasses` to provide some custom filter. It *only* declares that filter, as it *only* provides this filter. The application it depends on JAXB, which so far was mandatory on JAX-RS. So it *does not* register it explicitly, but just relies on it "being there somehow". JAXB is not mandatory anymore, so the depoloyer (not the application programmer) adds an `Extension` module which shall get auto-loaded to provide it. This case makes clear, that it **must not** be up to the **application** to explicitly provide JAXB, as it is the **deployer's** choice. Also the case makes clear, that implementing `getClasses` **must not** disable auto-loading of the extension, as it would effectively **break** the application. So am I strictly **pro `Extension`**, strictly **-1 for disabling auto-loading of `Extension`, and +-0 for having some other opt-out mechanism.

Reply:
(Maybe we should start discussing *separate* Classloaders, one for application-provided providers, and one for bundled *`Extension`s*?

Reply:
Suppose we have `foo.jar`, which contains `Feature` X as a service. Then with this PR in its current state: if `foo.jar` is deployed on an internal server path (e.g. `glassfish/modules` in Payara), it will be picked up by the JAX-RS implementation. This is currently not possible to do in a portable way and, as I understand it, the main use case. If `foo.jar` is part of a WAR, behavior matches the [existing discovery mechanism](https://jakarta.ee/specifications/restful-ws/3.0/jakarta-restful-ws-spec-3.0.html#servlet) and therefore guarantees backward compatibility.

@mkarg I didn't follow the history there, but the current construction w.r.t. JAXB is unconventional to say the least. Anyway, since the related API is deprecated, I'd expect for the application developer to migrate away from it, and for the deployer to add any missing JARs on an internal server path in the meantime

Reply:
There is nothing unconventional about this use of JAXB. It is how Jersey does it, and Jersey is anything but unconventional.

Reply:
(Off-topic: I was referring to the fact that the JAXB clause in the module descriptor is `requires static`, even though it's still depended on by exported API)

Reply:
JAXB was just an arbitrary example and was not meant literally. Just pick *any* Jersey module to get the point.

Reply:
I thought I got the point and replied to it, but let me try again: either an application developer develops against the spec, and bundles everything that isn't required by the spec in their application. Or they agree with their server administators/deployers about a set of additional features that will be provided to their application. In the latter case, if the JAX-RS implementation doesn't provide all of the agreed-upon features, the server administrator/deployer simply adds the required features on an internal server path, as is supported by this PR, and that is that.

Reply:
And exactly this scenario is what I try to solve with standardized `Extension` modules: You are not bound to the vendor-specific way of supplying the modules "in some internal server path" but can rely upon the JAX-RS standard, hence it will work exactly the same with all implementations, even in the cloud where you cannot talk to the administrator at all.

Reply:
But in that case, the application developer shouldn't make any assumptions about what is and isn't available, and just develop against the spec, no? I must be missing something, so will you please give a concrete example, where it's neither acceptable for the application developer to bundle the `Feature`, nor possible for the deployer to add it to the server? (PS: I've edited the comment above to remove "internal server path", because it's irrelevant: every Jakarta EE server has its own way to extend it, but if you want to standardize that, you should do so at the Platform level, not for each specification individually)

Reply:
Think of an ISV that provides an EAR to the public. The ISV cannot know whether or not your application server provides support for JAXB (again, just an example), as there are cloud providers on the market that provide strict JAX-RS compliant service offerings, and others that provide several optional modules ontop. So the deployer (= customer of the ISV) can choose to either deploy to a cloud service officially supporting JAXB (= using a highly optimized and perfectly tuned implementation of that cloud provider), or he can add a slow-but-cross-vendor-runnable JAXB module into the EAR (just drop it in *there*) that *he* (not the appliction vendor) selected from the free market.

Reply:
Thanks for the example. So IIUC, your point is that the deployer should be able to drop a JAR into the EAR, right? But if so, I still see several ways out. The ISV could e.g. ...
* use `ServiceLoader` inside the `Application` subclass to discover additional `Features` itself
* avoid adding an `Application` subclass that overrides `getClasses` and/or `getSingletons`
* provide different builds of its EAR for different deployment environments
* simply state `our EAR is only certified against these Jakarta EE implementations` (just as it could state that it requires Java SE 11 e.g.)
* provide Docker images where everything's already bundled up and the application server has already been optimized for their EAR

Reply:
* You are mixing application vendor and runtime vendor here. My target is to have a clear separation of the choice of the application vendor ("I need JAXB support") from the choice of the deployer (= customer) ("I fulfil this wish using this cross-vendor runtime module from Apache). The idea of Java EE ever was to have this clear separation, and my invention of the `Extension` interface reflects exactly this separation.
* It is (since ever) the choice of the application vendor whether to implementation `Application` or not. EARs are not provided by vendors, they are provided by deployers.
* Having different pre-bundled EARs completely undermines the idea of vendor-independence and WORA that is fundamental to JAX-RS: If we agree to that, we can drop the idea of common standard at all but just use native Jersey API. * JAX-RS is a standalone API which must work the same within or without Jakarta EE, as there are JAX-RS compliant cloud offerings which never will be Jakarta EE compliant by intention.
* Forcing people into usage of Docker mixes up vendor-decisions, deployer-decisions and provider-decisions and completely undermines the idea of a common Java-based standard, hence is out of scope of JAX-RS. We cannot answer the request to standardize the existing vendor-specific flags that identifies JAX-RS runtime modules by saying "use Docker" as there are people out there not using Docker at all.

To sum up, you proposed a lot of workarounds, but none of your answers solves the actual request: Standardizing the existing vendor-specific runtime module identifiers in a cross-vendor JAX-RS standard way.

Reply:
* all I'm saying is: if desired, the ISV can accommodate deployers that want to inject additional `Features` in their EAR, despite the presence of the ISV's `Application` subclass
* in your example you say "an ISV that provides an EAR" and "the deployer (= customer of the ISV)", but now you say "EARs are not provided by vendors, they are provided by deployers."? Anyway, yes, it's up to the ISV to decide, and I gave an example of how to accommodate deployers in both situations
* what I meant is: an EAR with `jaxb.jar` (which has a service provider for `Feature` that provides JAXB support), and an EAR without `jaxb.jar`. The code in `jaxb.jar` itself would be pure JAX-RS
* as an ISV, I can state whatever I want, so I don't get your point
* I gave 4 alternative options, so I don't understand how that is forcing anything. Moreover, I don't get the point: in this scenario, the ISV sells a "solution", and is both application developer & deployer at the same time. And when building its Docker image, they can use the functionality provided by this PR to extend the server runtime in a standardized way, independently of their EAR

> To sum up, you proposed a lot of workarounds, but none of your answers solves the actual request: Standardizing the existing vendor-specific runtime module identifiers in a cross-vendor JAX-RS standard way.

Maybe none of my answers do, but I believe my PR does.

Reply:
This discussion is getting a bit hard to follow, but I want to comment on a few aspects:

> > use ServiceLoader inside the Application subclass to discover additional Features itself

> You are mixing application vendor and runtime vendor here. My target is to have a clear separation of the choice of the application vendor ("I need JAXB support") from the choice of the deployer (= customer) ("I fulfil this wish using this cross-vendor runtime module from Apache")

My thoughts for this concrete example: As JAXB isn't required for Jakarta EE anymore, an ISV should not have JAXB as a requirement. I agree with @anthonyvdotbe that in this case the ISV should simply provide a fallback for environments where JAXB is not available, which should be easy by registering a provider with a low priority.

> EARs are not provided by vendors, they are provided by deployers.

IMO maybe in an ideal world, but not in the real world. 

> * Having different pre-bundled EARs completely undermines the idea of vendor-independence and WORA that is fundamental to JAX-RS

I guess the real problem is that with optional features WORA doesn't really work. And to be honest: If a deployer has to "repackage" an ISV application to add support for something like JAXB, this isn't WORA anymore.

I would also like to comment on this reply by @mkarg:

> Look at this case: An application explicitly implements getClasses to provide some custom filter. It only declares that filter, as it only provides this filter. The application it depends on JAXB, which so far was mandatory on JAX-RS. So it does not register it explicitly, but just relies on it "being there somehow". JAXB is not mandatory anymore, so the depoloyer (not the application programmer) adds an Extension module which shall get auto-loaded to provide it. This case makes clear, that it must not be up to the application to explicitly provide JAXB, as it is the deployer's choice. 

I don't agree to the *must not*. Both ways should be possible. Either the deployer adds JAXB support (by adding it to the EAR or whatever) OR the application could include JAXB support itself (either by adding some kind of fallback provider OR by registering a provider with higher priority that is always preferred.

> Also the case makes clear, that implementing getClasses must not disable auto-loading of the extension, as it would effectively break the application. 

Please note that according to the current proposal implementing `getClasses` will only disable auto-discovery of extensions in the application's classloader. So according to my understanding extensions installed in the app server and extensions provided via EARs are still loaded.

> So am I strictly pro Extension, strictly **-1 for disabling auto-loading of Extension, and +-0 for having some other opt-out mechanism.

I still don't understand the benefit of having this additional `Extension` interface TBH. And I still don't see a need for distinguishing between runtime extensions and other extensions. I still see the mechanism we are discussing here primarily for portable libraries like Soteria, Mojarra, Payara JWT and Krazo.

Reply:
In the real world, there *are* scenarios where application vendor, deployer, and vendor of the module are *distrinct* entities that solely trust on the fact that there is a **standard** way to identify and deploy runtime modules, and that way is part of the JAX-RS standard, not outside of it, not vendor-specific, not application-bound. The need for a separate interface is that in future we *might* want to have methods in that feature, allowing runtimes to call on these modules (like for initialization purposed etc). And yes, this *mostly* is needed for portable modules, certainly, as this is *the reason* **why I invented `Extension` interface**.

Reply:
@chkal Again, please note that "JAXB" was just an example, please replace it by "any runtime module". I already answered this several times.

Reply:
> In the real world, there _are_ scenarios where application vendor, deployer, and vendor of the module are _distrinct_ entities that solely trust on the fact that there is a **standard** way to identify and deploy runtime modules, and that way is part of the JAX-RS standard, not outside of it, not vendor-specific, not application-bound.

Agreed. And if we decide to merge this PR, it should support this scenario. Although I still don't see a real difference between a "runtime module" and a "portable library". In my view it is essentially the same. But maybe I'm missing something.

> The need for a separate interface is that in future we _might_ want to have methods in that feature, allowing runtimes to call on these modules (like for initialization purposed etc). And yes, this _mostly_ is needed for portable modules, certainly, as this is _the reason_ **why I invented `Extension` interface**.

So you say there is no need for the `Extension` interface *today* and you would like to keep it, so we can add methods later? I agree that we *may* have a need for additional methods later on, but wouldn't it make more sense to add such methods to the `Feature`/`DynamicFeature` interfaces instead? Basically a `Feature` is a set of providers. And as mentioned before, I don't see a need to distinguish between "runtime modules" and "portable libraries" which are both just sets of providers. That's what I like about this proposal: An extension is just a `Feature` (or `DynamicFeature`) which is loaded automatically and registers a list of providers.

Reply:
Yes you are missing something indeed: Runtime modules might exist that provide implementations *not intended* for applications. For example, a runtime module might exist *in future* that implements parts of RuntimeDelegate (i. e. RuntimeDelegate uses customizable factories). Such a factory will never be part of an application neither will it be a feature, and such a module must be loaded *before* the runtime is even trying to scan the application. So there is a clear separation between runtime modules and application components *in theory*. Yes, we did not plan to cover these cases *now*, but it makes sense to start *somewhere*, so we should keep the API open for such enhancements that go much deeper into that separation *in future*. As a result, we *must* keep the interface, and we *must* keep the separation between extensions (runtime modules) and features (application components). Extensions *can* be features (and as of today possibly will be in *many* cases), but they *not necessarily* are.

This is why I always said, extensions are something *different* to application components providers since I invented the interface months (years) ago, and this is why *this PR* is definitively **not** *an alterantive* to `Extension` as it does not even try to cover the sole use case I invented that interface for!

Reply:
> For example, a runtime module might exist in future that implements parts of RuntimeDelegate (i. e. RuntimeDelegate uses customizable factories). Such a factory will never be part of an application neither will it be a feature, and such a module must be loaded before the runtime is even trying to scan the application.

This is not a good example indeed. RuntimeDelegate implementation is strictly connected to the Jakarta REST implementation and as such, it should be the vendor's responsibility to get the mentioned RuntimeDelegate part onboard. 

I understand the API as an instrument for customers to use. For vendors, we shall define the requirements and demands, not Extensions they are free to use. It is more the customers who will use the Extension interface (or this counterproposal should it be approved) primarily, the vendors likely stay with the module loading mechanism they already have.

Reply:
I invented the `Extension` interface solely for implementation by vendors not users -- by runtime providers and by third party providers of cross-vendor runtime modules, actually, independent of what anybody *thinks* or *assumes*. Example: Application declares to use JSON-B, and deployer decides whether to use original Jersey-JSON-B-media-module or let Jersey use Johnzon-JSONB cross-vendor module. This is neither the application's choice not the runtime vendor's choice.